marconi-chain-index/src/Marconi/ChainIndex/Indexers/EpochStakepoolSize.hs

```diff
@@ -1,187 +1,502 @@
-  return $ Event (epochNo, M.fromList xs)
-
-queryPoolId :: SQL.Connection -> C.PoolId -> IO [(C.EpochNo, C.Lovelace)]
-queryPoolId c poolId = do
-  SQL.query c "SELECT epochNo, lovelace FROM stakepool_delegation WHERE poolId = ?" (SQL.Only poolId)
-
-queryAll :: SQL.Connection -> IO [Event]
-queryAll c = do
-  all' :: [(C.EpochNo, C.PoolId, C.Lovelace)] <- SQL.query_ c "SELECT epochNo, poolId, lovelace FROM stakepool_delegation ORDER BY epochNo ASC"
-  let
-    lastTwo (_, a, b) = (a, b)
-    result = all'
-      & groupBy ((==) `on` (\(e, _, _) -> e))
-      & map (\case xs@((e, _, _) : _) -> Just $ Event (e, M.fromList $ map lastTwo xs); _ -> Nothing)
-      & P.catMaybes
-  pure result
+        SQL.execute_ c "BEGIN"
+        forM_ (concatMap eventToEpochSPDRows $ filter epochSPDEventIsFirstEventOfEpoch eventsList) $ \row ->
+          SQL.execute c
+              [r|INSERT INTO epoch_spd
+                  ( epochNo
+                  , poolId
+                  , lovelace
+                  , slotNo
+                  , blockHeaderHash
+                  , blockNo
+                  ) VALUES (?, ?, ?, ?, ?, ?)|] row
+        SQL.execute_ c "COMMIT"
+
+        -- We store the LedgerState if one of following conditions hold:
+        --   * the LedgerState cannot be rollbacked and is the last of an epoch
+        --   * the LedgerState can be rollbacked
+        let writeLedgerState ledgerState (C.SlotNo slotNo) blockHeaderHash (C.BlockNo blockNo) isRollbackable = do
+                let fname = ledgerStateDirPath
+                        </> "ledgerState_"
+                         <> (if isRollbackable then "volatile_" else "")
+                         <> show slotNo
+                         <> "_"
+                         <> Text.unpack (C.serialiseToRawBytesHexText blockHeaderHash)
+                         <> "_"
+                         <> show blockNo
+                         <> ".bin"
+                -- TODO We should delete the file is the write operation was interrumpted by the
+                -- user. Tried using something like `onException`, but it doesn't run the cleanup
+                -- function. Not sure how to do the cleanup here without restoring doing it outside
+                -- the thread where this indexer is running.
+                BS.writeFile fname (CBOR.toLazyByteString $ encodeLedgerState ledgerState)
+        forM_ eventsList
+              $ \(EpochSPDEvent
+                    maybeLedgerState
+                    maybeEpochNo
+                    _
+                    slotNo
+                    blockHeaderHash
+                    blockNo
+                    chainTip
+                    isFirstEventOfEpoch) -> do
+            case (maybeEpochNo, maybeLedgerState) of
+              (Just _, Just ledgerState) -> do
+                  let isRollbackable = isBlockRollbackable securityParam blockNo chainTip
+                  when (isRollbackable || isFirstEventOfEpoch) $ do
```

Comment:
Right, so we write ledger state to disk when `isFirstEventOfEpoch` (this is what happens most when starting indexing from scratch, as none of the ledger states are volatile there), and then when we get close to tip, where I guess all the ledger states are volatile. Does this mean that we only get matching ledger state to roll back to when the rollback happens either to the previous slot or to the first slot for the epoch, and otherwise we error?

Reply:
I.e I guess rolling back only works in the case where we roll back just one slot?

Reply:
More precisely, we don't store any LedgerState from ByronEra, because they are no epochs in that era. So we only store the LedgerState starting at each epoch of ShellyEra and onwards.

> Does this mean that we only get matching ledger state to roll back to when the rollback happens either to the previous slot or to the first slot for the epoch, and otherwise we error?

No they work for _any_ rollbackable point. If the block is rollbackable, we store the LedgerState on disk. If securityParam is 2160, then we ultimately store 2160 volatile LedgerState on disk (and eventually delete the ones that are immutable). Therefore, we can always rollback to any point that we can rollback from.

Reply:
If the indexer catches up with the node such that it will start receiving volatile blocks then we write ledger state for each and every one of these? Is it so that given that in Cardano one block (slot) is one second then in reality we won't actually catch up because we'll be writing the ledger state to disk on every such block, and since writing ledger state to disk takes more than one second then effectively we'll always be 2160 slots behind?